libsolidity/formal/BMC.cpp

```diff
@@ -120,8 +120,19 @@ bool BMC::visit(ContractDefinition const& _contract)
 	return true;
 }
 
+void BMC::endVisit(ContractDefinition const& _contract)
+{
+	SMTEncoder::endVisit(_contract);
+}
+
 bool BMC::visit(FunctionDefinition const& _function)
 {
+	auto contract = dynamic_cast<ContractDefinition const*>(_function.scope());
+	solAssert(contract, "");
+	solAssert(m_currentContract, "");
+	if (contract != m_currentContract)
```

Comment:
Should be something like `cotract is not in inheritance hierarchy of m_currentContract`

Reply:
I was wondering why the tests actually passed even though this was wrong.
The reason is that `initializeStateVariables` only calls `createVariable` which does nothing if it was already created, so that's why it was safe.
I'll still push the fix to this soon.